sematic/ui/src/index.tsx

```diff
@@ -76,97 +64,26 @@ function App() {
-      setEnv(new Map());
-    }
-    fetchJSON({
-      url: "/api/v1/meta/env",
-      callback: (payload: EnvPayload) => {
-        setEnv(new Map(Object.entries(payload.env)));
-      },
-      apiKey: user?.api_key,
-    });
-  }, [user]);
-
-  const envContextValue = useMemo(() => env, [env]);
-
-  const onGoogleLoginSuccess = useCallback(
-    (credentialResponse: CredentialResponse) => {
-      fetchJSON({
-        url: "/login/google",
-        method: "POST",
-        body: {
-          token: credentialResponse.credential,
-        },
-        callback: (payload: GoogleLoginPayload) => {
-          setError(undefined);
-          localStorage.setItem("user", JSON.stringify(payload.user));
-          setUser(payload.user);
-        },
-        setError: setError,
-      });
-    },
-    []
-  );
-
-  return authenticate?.authenticate === false || user ? (
-    <UserContext.Provider value={userContextValue}>
-      <EnvContext.Provider value={envContextValue}>
-        <SnackBarProvider>
-          <Shell />
-        </SnackBarProvider>
-      </EnvContext.Provider>
-    </UserContext.Provider>
-  ) : (
-    <Paper
-      sx={{
-        width: 200,
-        p: 5,
-        textAlign: "center",
-        top: "50%",
-        left: "50%",
-        position: "absolute",
-        transform: "translateY(-50%) translateX(-50%)",
-      }}
-      variant="outlined"
-    >
-      <img
-        src={logo}
-        width="50px"
-        alt="Sematic logo"
-        style={{ marginBottom: "30px" }}
-      />
-      {authenticate === undefined ? <Loading isLoaded={false} /> : <></>}
-
-      {error ? <Alert severity="error">{error.message}</Alert> : <></>}
-
-      {!error &&
-      authenticate &&
-      authenticate.authenticate === true &&
-      authenticate.providers.GOOGLE_OAUTH_CLIENT_ID !== undefined ? (
-        <GoogleOAuthProvider
-          clientId={authenticate.providers.GOOGLE_OAUTH_CLIENT_ID}
-        >
-          <GoogleLogin
-            text="signin_with"
-            logo_alignment="center"
-            onSuccess={onGoogleLoginSuccess}
-            onError={() => {
-              setError(Error("Unauthorized user"));
-            }}
-          />
-        </GoogleOAuthProvider>
-      ) : (
-        <></>
-      )}
-    </Paper>
-  );
+  if (error || loading) {
+    return <Loading error={error} isLoaded={!loading} />;
+  }
+
+  return <AppContext.Provider value={appContextValue}>
+      <UserContext.Provider value={userContextValue}>
+        <EnvironmentProvider>
+          <SnackBarProvider>
+            <Outlet />
+          </SnackBarProvider>
+        </EnvironmentProvider>
+      </UserContext.Provider>
+    </AppContext.Provider>;
 }
 
-function Router() {
-  const router = createBrowserRouter(createRoutesFromElements(
-    <Route path="/" element={<App />}>
+const router = createBrowserRouter(
+  createRoutesFromElements(
+  <Route path="/" element={<App />}>
+    <Route path="/login" element={<GoogleLoginPage />} />
```

Comment:
Will this be generalizable in the future with other OAuth implementations?

Reply:
It should be. The `/login` page should be able to feature multiple Sign In methods depending on what the server returns.

Reply:
Yes, that is one goal of this refactoring. Apart from what @neutralino1 suggested, we can also do that in `shell.tsx`, where we redirect to different pages (`/login_goggle`, `/login_github`) based on the authentication metadata retrieved from the server.

Reply:
I think we always want a single login page, but with multiple login in buttons, as is usual in all apps.